### PR TITLE
feat: near subgraph

### PR DESCRIPTION
**Pull request by yomarion**
* Detection of NEAR native token payments (without conversion)

### CODE REVIEW

File: src/erc20FeeProxy.ts [hunk 1]
```diff
@@ -18,7 +18,7 @@ export function createPaymentForFeeProxy(
   payment.block = event.block.number.toI32();
   payment.timestamp = event.block.timestamp.toI32();
   payment.txHash = event.transaction.hash;
-  payment.gasUsed = event.transaction.gasUsed;
+  payment.gasUsed = event.transaction.gasLimit;
```
Comment (benjlevesque) [inline src/erc20FeeProxy.ts:21]: could you explain this change? gasLimit and gasUsed are not the same thing...

File: subgraph.template.yaml [hunk 2]
```diff
@@ -12,7 +12,7 @@ dataSources:
       startBlock: {{creationBlockNumber}}
     mapping:
       kind: ethereum/events
-      apiVersion: 0.0.4
+      apiVersion: 0.0.5
```
Comment (benjlevesque) [inline subgraph.template.yaml:15]: any reason not to use 0.0.6? 
also, the specVersion is now 0.0.4 , maybe we should update

not mandatory at all

File: schema.near.graphql [hunk 3]
```diff
@@ -0,0 +1,25 @@
+type NearPayment @entity {
```
Comment (benjlevesque) [inline schema.near.graphql:1]: could you add comments to know which fields are different from the other schema?

File: src/near-nativePayment.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  paymentReference: string,
+  to: string
+): void {
+  const receipt = receiptWithOutcome.receipt;
+  // TODO generateId not sufficient
+  let payment = new NearPayment(generateId(receipt.id));
+  payment.to = to;
+  payment.amount = amount;
+  payment.reference = paymentReference;
+  payment.contractAddress = "requestnetwork.near";
```
Comment (benjlevesque) [inline src/near-nativePayment.ts:92]: isn't this available in the receipt or something? it makes unusable for another network (testnet)

File: src/near-nativePayment.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+  payment.from = receipt.signerId;
+  payment.block = BigInt.fromI64(
+    receiptWithOutcome.block.header.height - 1
```
Comment (benjlevesque) [inline src/near-nativePayment.ts:96]: comment would be appreciated here

File: src/near-nativePayment.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    return;
+  }
+  if (!near.SuccessStatus.fromReceiptId(receipt.id)) {
+    return;
+  }
+
+  const outcome = receiptWithOutcome.outcome;
+
+  const functionCall = action.toFunctionCall();
+  if (functionCall.methodName == "transfer_with_reference") {
```
Comment (benjlevesque) [inline src/near-nativePayment.ts:42]: reverse the condition to avoid nesting?

File: src/near-nativePayment.ts [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  const actions = receipt.receipt.actions;
+  for (let i = 0; i < actions.length; i++) {
+    handleAction(actions[i], receipt);
+  }
+}
+
+function generateId(transactionHash: Bytes): string {
+  var id = transactionHash.toHex(); // + paymentReference.toHex().slice(2);
+  return crypto.keccak256(ByteArray.fromHexString(id)).toHex();
+}
```
Comment (benjlevesque) [inline src/near-nativePayment.ts:25]: if there is only one value, I don't think this is necessary, you can use the receiptId as the ID.

To consider for the ID:
- a transaction might contain several transfers
- a single payment reference might be present in several transactions

**Comment by yomarion** [comment]
> any reason not to use 0.0.6?
> also, the specVersion is now 0.0.4 , maybe we should update
> 
> not mandatory at all

@benjlevesque  I just followed the docs at: https://thegraph.com/docs/en/supported-networks/near/. We can try with 0.0.6 if you know some benefits

File: src/erc20FeeProxy.ts (hunk 1 above)
Comment (yomarion) [reply src/erc20FeeProxy.ts:21]: @benjlevesque should be done on another PR, not related with Near so I'll remove it now.

But the gasLimit is different from the gasUsed. Example of a tx where we retrieve the wrong number:
https://etherscan.io/tx/0xd0876ba91f2ef8c0d4687d782bb785d1e8f7478ccee485e1bd59df4acf9143f3

```
{
        "id": "0x1702404698caebd7a13b736fa7255d94dfaea53ccc866e143b7bf5228755aa4a",
        "contractAddress": "0x370de27fdb7d1ff1e1baa7d11c5820a324cf623c",
        "tokenAddress": "0xb64ef51c888972c908cfacf59b47c1afbc0ab8ac",
        "to": "0x8400b234e7b113686bd584af9b1041e5a233e754",
        "txHash": "0xd0876ba91f2ef8c0d4687d782bb785d1e8f7478ccee485e1bd59df4acf9143f3",
        "gasUsed": "100000"
      },
```

Comment (yomarion) [reply src/erc20FeeProxy.ts:21]: It's a typing issue from `graph-ts` that I quick-fixed with `gasLimit`

Comment (yomarion) [reply src/erc20FeeProxy.ts:21]: The graph does not expost receipt information like gasUsed at the moment so we mistakenly store gasLimit instead. So the type "fix" is actually mandatory to support Near, it does not change the behaviour, only makes the error more explicit. I'll add `FIXME` and refer to the issue below, ok for you @benjlevesque  ?

The Graph Issue here: https://github.com/graphprotocol/graph-node/issues/3218
Similar discussion here: https://github.com/graphprotocol/graph-node/issues/2619

Example here: 
0xd0876ba91f2ef8c0d4687d782bb785d1e8f7478ccee485e1bd59df4acf9143f3
gasLimit = 100k, gasUsage = 71k

And from our subgraph:
```
{
        "id": "0x1702404698caebd7a13b736fa7255d94dfaea53ccc866e143b7bf5228755aa4a",
        "contractAddress": "0x370de27fdb7d1ff1e1baa7d11c5820a324cf623c",
        "tokenAddress": "0xb64ef51c888972c908cfacf59b47c1afbc0ab8ac",
        "to": "0x8400b234e7b113686bd584af9b1041e5a233e754",
        "txHash": "0xd0876ba91f2ef8c0d4687d782bb785d1e8f7478ccee485e1bd59df4acf9143f3",
        "gasUsed": "100000"
      },
```

Comment (benjlevesque) [reply src/erc20FeeProxy.ts:21]: I suggest we store 0 instead?

Comment (yomarion) [reply src/erc20FeeProxy.ts:22]: We can also do that yes but `gasLimit` is closer to the true answer so it depends on the usage we have. I'll check the impact and let you know.

Comment (benjlevesque) [reply src/erc20FeeProxy.ts:22]: it's not used because of the issue you mentionned at the moment

Comment (yomarion) [reply src/erc20FeeProxy.ts]: OK I just set it to 0

**Review by benjlevesque — APPROVED** [review]

File: README.md [hunk 8]
```diff
@@ -2,10 +2,15 @@
 
 This repo contains the code and configuration for Request Payment subgraphs:
 
-- Mainnet (coming soon)
+- [Mainnet](https://thegraph.com/hosted-service/subgraph/requestnetwork/request-payments-mainnet)
 - [Rinkeby](https://thegraph.com/explorer/subgraph/requestnetwork/request-payments-rinkeby)
 - [Matic](https://thegraph.com/explorer/subgraph/requestnetwork/request-payments-matic)
```
Comment (benjlevesque) [inline README.md:7]: [Polygon (Matic)]

File: README.md [hunk 9]
```diff
@@ -2,10 +2,15 @@
...
 - [Celo](https://thegraph.com/explorer/subgraph/requestnetwork/request-payments-celo)
+- [BSC](https://thegraph.com/hosted-service/subgraph/requestnetwork/request-payments-bsc)
+- [xDai (Gnosis Chain)](https://thegraph.com/hosted-service/subgraph/requestnetwork/request-payments-xdai)
```
Comment (benjlevesque) [inline README.md:10]:
```suggestion
- [Gnosis Chain (xDai)](https://thegraph.com/hosted-service/subgraph/requestnetwork/request-payments-xdai)
```

**Review by benjlevesque — APPROVED** [review]

**Pull request merged**